appcelerator/titanium_mobile
# [TIMOB-26306] Android: Improved Titanium SDK (titanium_mobile) ant/ndk builds

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-26306
- https://jira.appcelerator.org/browse/TIMOB-25932

**Summary:**
- [TIMOB-26306](https://jira.appcelerator.org/browse/TIMOB-26306) Can now do fast parallel execution `-j8` NDK builds on Windows.
- [TIMOB-25932](https://jira.appcelerator.org/browse/TIMOB-25932) Can now build with JDK 10.
  * Removed "javah" tool usage. Tool was removed from JDK 10, which caused build failures.
  * Removed usage of "javah" generated header files. Not needed. Functions on C++ side are already exported via `JNIEXPORT`.
- Refactored Android make files.
  * Moved code generation parts to "ant" as a pre-build step. _(Fixes dependency order and file locking issues.)_
  * Removed files "common.mk" and "genSources.mk".
  * Added missing `APP_PLATFORM` setting. _(We were wrongly using `TARGET_PLATFORM` instead.)_
- Modified gperf and python code generators to create `*.h` files instead of `*.cpp` files.
  * Greatly simplifies make file configuration.
  * Dependency order between `*.cpp` no longer needs to be defined. _("KrollBindings.cpp" was `#include`ing these `*.cpp` files.)_
  * Simplifies `*.cpp` file listing in make file. _(No longer have to omit generated `*.cpp` files.)_
- Fixed majority of Android Java/C/C++ compiler warnings.
- Modified "javac" to ignore Java 1.7 bootstrap warning via "-Xlint:-options".
  * Only way to not ignore it is to reference Java 7 "rt.jar" which is not bundled with JDK 8 or higher.
  * Resolves `warning: [options] bootstrap class path not set in conjunction with -source 1.7`
- Modified NDK "clean" to delete the x86 intermediates/binaries.
  * Was missing `BUILD_X86` environment variable setting.
- Updated Titanium project files to use "min" API Level 16.
  * Note that not all files were updated when we last raised it. Particularly the modules and ndk side.
  * Fixes warnings with NDK build files.
- I've tested this PR with both Android NDK `r12b` and `r17b`.

**Test:**
Do the below on both Mac and Windows.
1. Install JDK 10 if not done already.
2. Open the Terminal or Command Prompt.
3. `cd` to the `titanium_mobile/build` directory.
4. Enter: `node scons.js cleanbuild`
5. Verify that the build finishes successfully.
6. Search the Terminal for "warning". Verify that no warnings appear.
7. Build and run a Titanium app for Android using the "7.4.0" version you just built.
8. Verify that the app launches successfully on Android.
